src/goto-programs/restrict_function_pointers.cpp

```diff
@@ -0,0 +1,415 @@
+/*******************************************************************\
+
+Module: GOTO Program Utilities
+
+Author: Diffblue Ltd.
+
+\*******************************************************************/
+
+#include "restrict_function_pointers.h"
+
+#include <ansi-c/expr2c.h>
+#include <json/json_parser.h>
+#include <util/expr_iterator.h>
+#include <util/string_utils.h>
+
+#include <algorithm>
+#include <fstream>
+#include <iostream>
+
+namespace
+{
+void typecheck_function_pointer_restrictions(
+  const goto_modelt &goto_model,
+  const function_pointer_restrictionst &restrictions)
```

Comment:
:green_heart: awesome checks - this is exactly the kind of things that would be super frustrating from user perspective to only find out how letting the analysis run!
Be great to have a unit test for each of the error checks here